### PR TITLE
Use runtime.CallersFrames to create StackFrames

### LINKED ISSUE #44: errors: Tests fail
Running `go test` fails (go 1.7.4), because the stack formatting never matches.

### PR DESCRIPTION
**Pull request by aduong**
## Goal

Stack traces from `errors.Error` are incomplete and miss functions. In particular, inlined functions are missing.

Additionally, the tests for `errors` were failing.

Also fixes #44 

## Design

The change is to user `runtime.CallersFrames` to translate program counters to symbolic information as recommended by the golang docs.

### Added
`errors.NewStackFrameFromRuntime`

### Removed
None

### Changed
`errors.Error.StackFrame`

## Tests

Unit tests were fixed to fail for the right reasons (stack frame content, not formatting). All unit tests were run and passed.

### Outstanding Questions

`errors.StackFrame.Func()` still relies on `runtime.FuncForPC` which is less than ideal. Within the library, it is primarily used on constructor of `StackFrame`s. One way to give the right `Func` would be to save the `Func` on the `StackFrame` itself on construction from a `runtime.Frame`.

## Review
For the submitter, initial self-review:

- [X] Commented on code changes inline explain the reasoning behind the approach
- [X] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [X] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [X] Initial review of the intended approach, not yet feature complete
  - [X] Structural review of the classes, functions, and properties modified
  - [X] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

**Comment by bengourley** [comment]
Hi @aduong, thanks for the PR. Due to the scope of this it'll take us a while to get it reviewed. I can't give you an ETA at this point, but we have a lot of other higher priority things going on at the moment. Thanks for your patience!

File: errors/stackframe.go [hunk 1]
```diff
@@ -31,7 +30,18 @@ func NewStackFrame(pc uintptr) (frame StackFrame) {
 	// and we want to show the line that corresponds to the function call
 	frame.File, frame.LineNumber = frame.Func().FileLine(pc - 1)
 	return
+}
 
+// NewStackFrameFromRuntime populates a stack frame object from a runtime.Frame object.
+func NewStackFrameFromRuntime(frame runtime.Frame) StackFrame {
+	pkg, name := packageAndName(frame.Func)
```
Comment (mkobetic) [inline errors/stackframe.go]: It seems that frame.Func can still be nil sometimes. Might be worth attempting to salvage something from frame.Function in that case.

Comment (aduong) [reply errors/stackframe.go]: Yes, that seems reasonable. I've pushed a commit that does this.

**Comment by aduong** [comment]
> Hi @aduong, thanks for the PR. Due to the scope of this it'll take us a while to get it reviewed. I can't give you an ETA at this point, but we have a lot of other higher priority things going on at the moment. Thanks for your patience!

@bengourley, thanks for the prompt response. If it helps, I can split this change set so that the failing tests are separate set. As it turns out, the error_test suite was already being skipped because it was platform-dependent. Part of this change set makes those tests platform independent.

**Comment by aduong** [comment]
Sadly, this doesn't seem the team had the capacity to review this and this is probably too outdated. My team has been able to work around the issue. Perhaps this has been fixed in later versions.

**Comment by mathewbyrne** [comment]
Is there any chance of reviving this PR?

We're running into an issue where we have an incorrect stack trace being generated for a panic that looks to be corrected by this change.  It's very difficult to debug further into the runtime, but seems like common sense to follow the guidance of the Go documentation:

> To translate these PCs into symbolic information such as function names and line numbers, use CallersFrames. CallersFrames accounts for inlined functions and adjusts the return program counters into call program counters.

We may consider forking our own client to get this fix in, but if it this could be merged here that would be much preferable.

**Comment by xander-jones** [comment]
Hey @mathewbyrne, seen your message on https://github.com/bugsnag/bugsnag-go/pull/114 so we'll continue the conversation there.
